testcases/open_posix_testsuite/conformance/interfaces/strcpy/1-1.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * This file is licensed under the GPL license. For the full content
+ * of this license, see the COPYING file at the top level of this
+ * source tree.
+ */
+
+/*
+ * assertion:
+ *	The function will copy the string pointed to by s1 (including the
+ *	terminating NUL character) into the array pointed to by s2. And
+ *	it will return s2.
+ *
+ * method:
+ *	-Generate sample string s1 of variable lengths in a loop.
+ *	-Use strcpy() to copy s1 into sample string s2 in each iteration.
+ *	-Compare both strings(s1 and s2).
+ *	-Also compare returned pointer with s2.
+*/
+
+#define _XOPEN_SOURCE 700
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <time.h>
+#include "posixtest.h"
+
+#define TNAME "strcpy/1-1.c"
+
+char *random_string(int len) {
+	char *input_string = "01234567890\n\t\r\v\fabcdefghijklmnopqrstuvwxyz~!#$%^&*()_+ABCDEF"
+							"GHIJKLMNOPQRSTUVWXYZ";
+	size_t string_len = strlen(input_string);
+	short key = 0;
+	char *output_string = NULL;
+	srand(time(NULL) * len);
+	output_string = malloc(sizeof(char) * (len + 1));
+	for (int i = 0; i < len; i++) {
+		key = rand() % string_len;
+		output_string[i] = input_string[key];
+	}
+	output_string[len] = '\0';
+	return output_string;
+}
+
+int main(void)
+{
+	char *ret_str = NULL;
+
+	for (int i = 0; i < 50000 ; i = i + 2000) {
+		char *sample_str_1;
+		char *sample_str_2 = malloc(sizeof(char) * (i + 1));
+
+		sample_str_1 = random_string(i);
+		ret_str = strcpy(sample_str_2, sample_str_1);
+
+		if (strcmp(sample_str_1, sample_str_2) == 0 && ret_str == sample_str_2) {
+			printf(TNAME " Strcpy worked for string length %d\n", i);
+			free(sample_str_1);
+			free(sample_str_2);
+		} else if (strcmp(sample_str_1, sample_str_2) != 0) {
+			printf(TNAME " Test Failed, string copy failed. "
+						"Expected string: %s, But got: %s\n"
+						, sample_str_1, sample_str_2);
+			free(sample_str_1);
+			free(sample_str_2);
+			exit(PTS_FAIL);
+		} else {
+			printf(TNAME " Test Failed, return is not as expected. "
+						"Expected return: %p, But obtained: %p\n"
+						, sample_str_2, ret_str);
+			free(sample_str_1);
+			free(sample_str_2);
+			exit(PTS_FAIL);
+		}
+	}
+	printf(TNAME " Test Passed, string copied successfully\n");
+	exit(PTS_PASS);
```

Comment:
This is way too ugly, we really don't have to free() the strings in each if/else branch.

We can safely do that once at the end of the iteration, since in a case of failure we call exit(PTS_FAIL) anyway, which will exit the test and frees all program memory.

Reply:
Again i think it is good practice to free dynamically allocated memory before exiting from program.

Reply:
It sounds like you got some ideas on "good practice" in programming on a university course. Let me tell you something, these courses are often taught by people who wrote papers on "good practices" rather than people who maintain code for living.

If you haven't done so already have a look at what Linus and Greg wrote about C kernel programming:

https://www.kernel.org/doc/html/v4.10/process/coding-style.html
http://www.kroah.com/linux/talks/ols_2002_kernel_codingstyle_talk/html/mgp00001.html

And no, it does not really matter if you free() the memory or not right before your program exits main(), that's like cleaning up the carpet in a building that is demolished right after that. But even if you want to do that the code you wrote is ugly since it repeats the same piece of code over and over.